modules/generic-scala3/src/main/scala/pureconfig/generic/scala3/HintsAwareConfigReaderDerivation.scala

```diff
@@ -0,0 +1,31 @@
+package pureconfig
+package generic
+package scala3
+
+import scala.compiletime._
+import scala.deriving.Mirror
+
+trait HintsAwareConfigReaderDerivation
+    extends HintsAwareCoproductConfigReaderDerivation,
+      HintsAwareProductConfigReaderDerivation {
+  inline def deriveReader[A](using m: Mirror.Of[A], ph: ProductHint[A], cph: CoproductHint[A]): ConfigReader[A] =
+    inline m match {
+      case pm: Mirror.ProductOf[A] => deriveProductReader[A](using pm, ph)
+      case sm: Mirror.SumOf[A] => deriveSumReader[A](using sm, cph)
+    }
+
+  protected inline def summonConfigReader[A]: ConfigReader[A] =
+    summonFrom {
+      case reader: ConfigReader[A] => reader
+      case m: Mirror.Of[A] =>
+        deriveReader[A](using m, summonInline[ProductHint[A]], summonInline[CoproductHint[A]])
+    }
+}
+
+object HintsAwareConfigReaderDerivation extends HintsAwareConfigReaderDerivation {
+  object syntax {
```

Comment:
Yeah, agreed, I missed this one.

Reply:
I removed it. It's easier to bring something back, rather than trying to work around it in already published artifact. 😄 

In general however, I disagree with the notion that `derives` + implicit hints is harder to reason about than `deriveReader` + implicit hints. I could be missing something, but to me the first version is just a nicer syntax for the same thing. 
I suspect people will be adding `extension (c: ConfigReader.type) inline def derived[A](using Mirror.Of[A]): ConfigReader[A] = deriveReader[A]` to their codebases anyway.

One thing that is definitely hard to reason about is that we rely on _default_ implicit config. This is how it works in `pureconfig-generic` so we shouldn't change that for this module, but personally I much prefer the the model from new Scala 3 configured derivation in circe-core where you have to explicitly state the config.
But the `implicit def` that pureconfig relies on will at some point be removed in Scala 3, so it might force the users to state the config with `import defaults.given` anyway. 😅

Reply:
> One thing that is definitely hard to reason about is that we rely on _default_ implicit config. This is how it works in `pureconfig-generic` so we shouldn't change that for this module, but personally I much prefer the model from new Scala 3 configured derivation in circe-core where you have to explicitly state the config.

Circe makes it more obvious that some configuration is taking place by using the `ConfiguredEncoder`, `ConfiguredDecoder` and `ConfiguredCodec` types, but I would prefer a simpler model for PureConfig. For example, let's say that we have the following coproduct definitions:

```scala
sealed trait Foo
case class Bar(a: Int, b: String) extends Foo
case class Baz(c: Double, e: Boolean) extends Foo
```

If we want to add default `Codec` instances for each of these types we can do the following:

```scala
sealed trait Foo derives Codec
case class Bar(a: Int, b: String) extends Foo derives Codec
case class Baz(c: Double, e: Boolean) extends Foo derives Codec
```

Or, if we want to configure the way serialization/deserialization works, use a `ConfiguredCodec` instead and have a `Configuration` in scope:

```scala
given conf: Configuration = Configuration.default

sealed trait Foo derives ConfiguredCodec
case class Bar(a: Int, b: String) extends Foo derives ConfiguredCodec
case class Baz(c: Double, e: Boolean) extends Foo derives ConfiguredCodec
```

However, if we want to have a different configuration in just one of the types we would have to do something like the following:

```scala
given conf: Configuration = Configuration.default

sealed trait Foo derives ConfiguredCodec

case class Bar(a: Int, b: String) extends Foo

object Bar {
  given codec: ConfiguredCodec[Bar] =
    ConfiguredCodec.derived(using Configuration.default.withPascalCaseMemberNames)
}

case class Baz(c: Double, e: Boolean) extends Foo derives ConfiguredCodec
```

The two main things that I don't like with this approach is that (1) we need to be aware of more types and (2) we might not always be able to use the `derives` clause when configuration is to take place.

In PureConfig, I would prefer reducing the number of types we would need to know to `ConfigReader` and `ConfigWriter`, and have this process simplified to the point where `derives` would only produce default instances and explicit given instances were to be defined whenever some configuration was to take place.